File: backend/app/routers/requests.py
from fastapi import APIRouter, Depends, HTTPException, Response, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.request import Request
from app.schemas.request import RequestCreate, RequestRead

router = APIRouter(prefix="/requests", tags=["requests"])


@router.get("", response_model=list[RequestRead])
def get_requests(db: Session = Depends(get_db)):
    return db.query(Request).order_by(Request.id).all()


@router.get("/{request_id}", response_model=RequestRead)
def get_request(request_id: int, db: Session = Depends(get_db)):
    req = db.query(Request).filter(Request.id == request_id).first()
    if not req:
        raise HTTPException(status_code=404, detail="Request not found")
    return req


@router.post("", response_model=RequestRead, status_code=status.HTTP_201_CREATED)
def create_request(payload: RequestCreate, db: Session = Depends(get_db)):
    req = Request(**payload.model_dump())
    db.add(req)
    db.commit()
    db.refresh(req)
    return req


@router.put("/{request_id}", response_model=RequestRead)
def update_request(request_id: int, payload: RequestCreate, db: Session = Depends(get_db)):
    req = db.query(Request).filter(Request.id == request_id).first()
    if not req:
        raise HTTPException(status_code=404, detail="Request not found")

    data = payload.model_dump()
    for field, value in data.items():
        setattr(req, field, value)

    db.commit()
    db.refresh(req)
    return req


@router.delete("/{request_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_request(request_id: int, db: Session = Depends(get_db)):
    req = db.query(Request).filter(Request.id == request_id).first()
    if not req:
        raise HTTPException(status_code=404, detail="Request not found")

    db.delete(req)
    db.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/api/requests.js
import apiClient from "./client";

export async function getRequests() {
  const response = await apiClient.get("/requests/");
  return response.data;
}

export async function getMyRequests() {
  const response = await apiClient.get("/requests/my");
  return response.data;
}

export async function getRequest(requestId) {
  const response = await apiClient.get(`/requests/${requestId}`);
  return response.data;
}

export async function createRequest(requestData) {
  const response = await apiClient.post("/requests/", requestData);
  return response.data;
}

export async function updateRequest(requestId, requestData) {
  const response = await apiClient.put(`/requests/${requestId}`, requestData);
  return response.data;
}

export async function deleteRequest(requestId) {
  const response = await apiClient.delete(`/requests/${requestId}`);
  return response.data;
}

